zebra-consensus/src/checkpoint.rs

```diff
@@ -155,4 +152,231 @@ where
     }
 }
 
-// TODO(teor): tests
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use color_eyre::Report;
+    use eyre::{bail, ensure, eyre};
+    use tower::{util::ServiceExt, Service};
+
+    use zebra_chain::serialization::ZcashDeserialize;
+
+    #[tokio::test]
+    #[spandoc::spandoc]
+    async fn checkpoint_single_item_list() -> Result<(), Report> {
+        let block0 =
+            Arc::<Block>::zcash_deserialize(&zebra_test_vectors::BLOCK_MAINNET_GENESIS_BYTES[..])?;
+        let hash0: BlockHeaderHash = block0.as_ref().into();
+
+        // Make a checkpoint list containing only the genesis block
+        let genesis_checkpoint_list: HashMap<BlockHeight, BlockHeaderHash> =
+            [(block0.coinbase_height().unwrap(), hash0)]
+                .iter()
+                .cloned()
+                .collect();
+
+        let mut state_service = Box::new(zebra_state::in_memory::init());
+        let mut checkpoint_verifier = super::init(state_service.clone(), genesis_checkpoint_list);
+
+        // Verify block 0
+        let verify_response = checkpoint_verifier
+            .ready_and()
+            .await
+            .map_err(|e| eyre!(e))?
+            .call(block0.clone())
+            .await
+            .map_err(|e| eyre!(e))?;
+
+        ensure!(
+            verify_response == hash0,
+            "unexpected response kind: {:?}",
+            verify_response
+        );
+
+        let state_response = state_service
+            .ready_and()
+            .await
+            .map_err(|e| eyre!(e))?
+            .call(zebra_state::Request::GetBlock { hash: hash0 })
+            .await
+            .map_err(|e| eyre!(e))?;
+
+        match state_response {
+            zebra_state::Response::Block {
+                block: returned_block,
+            } => assert_eq!(block0, returned_block),
```

Comment:
I think you might be able to get away with asserting equality of the response type without the match with something like

`assert_eq!(zebra_state::Response::Block { block: block0 }, state_response);`

though this might require adding some `Eq` or `PartialEq` derives to `zebra_state::Response`, which is _probably_ fine but we should maybe confer with @hdevalence and @dconnolly before making that call.

Reply:
My only concern with adding `Eq` or `PartialEq` derives is that if we ever wanted to return non-comparable data we'd have to rip out all the derives or else add some custom shim to patch the gap.

Reply:
In this case I think that a clean way to write this code would be:
```
if let zebra_state::Response::Block { block } = rsp {
   assert_eq!(block, expected_block);
} else {
    bail!("unexpected response: {:?}", rsp);
}
```
and then there's no need to have `Eq` or to have deeply nested `match` statements.

Reply:
> My only concern with adding `Eq` or `PartialEq` derives is that if we ever wanted to return non-comparable data we'd have to rip out all the derives or else add some custom shim to patch the gap.

This seems like a reason to only impl `ParitalEq` which can represent non-comparable data.

Reply:
Fixed in #538 in e2e47d11abbd2b117b34dc4da5aae3ec3f5f4a6c for verify/block (but not checkpoint), using `if let` destructuring.

Reply:
Fixed in #538 in ab67803dca91167ea12ac5979abc13f177bb98de for checkpoint, using `if let` destructuring.